### DISCUSSION TITLE
About Warm-Up and Learning

### DISCUSSION THREAD

**Discussion by hrdrq** [EA31337/EA31337#326] [Q&A] [answered]
I've read your thread in Forex Factory. There are some descriptions about "warm-up", but in the latest code, `InitNoOfDaysToWarmUp` is not existing. Is EA not running "warm-up"?
I'm testing Lite with default params on EURUSD, when I started to run it at the beginning of this month, it almost only opened one order every 2~4 days.
But recently after 3 weeks, it's opening orders more often, even several orders within one day.
It looks like it finished "warm-up" to me. Is it just a coincidence that recent market is better for the EA?

You also mentioned that each run is unique and EA learns about market conditions.
How does "learning" works? After reading `EA.mqh`, `Strategy.mqh`, and `Trade.mqh` in the latest code, I didn't find the learning part. Could you share some hints? Thank you

**Comment by kenorb** [comment, marked as answer]
### Initial warm period

_Initial warm period_ param is implemented in v1 EAs (amount of initial days to lower the lot size), but not present in latest v2 due to logic refactor. In v2 the lot size calculation has been improved so there is no need due to lower risk (see: CalcLotSize). Secondly, the main reason of removal of this param is explained in 'logic refactor' section.

> It almost only opened one order every 2~4 days.

This isn't related to the warm period. It's just the combination of selected strategies. For example, Alligator strategy, the way how it's configured, it doesn't trade very often. It also depends on which symbol you're running it. Based on the backtest for [Lite v2.009 in 2020 on EURUSD](https://github.com/EA31337/EA31337/blob/v2.009/docker/backtest/Lite/all-yearly/2020/_results/EA31337-Lite_2020_10000USD_spread10.txt) it opens total 685 trades, so it's around 2 trades a day.

### Learning

In v1 (e.g. v1.080), the "learning" refers to set of params such as _Strategy boosting_, _Per strategy_ and _Smart queue_ parameters, where the longer EA runs, it can gather more statistics based on the current market conditions and strategy performance (such as profit factor, consecutive wins) and changes certain priorities which can influence the overall trading behaviour based on these criteria.

For example:

- ProfitFactor(Min/Max)ToTrade - EA stops strategy from trading when certain profit factor (PF) has been reached. So based on PF, it can learn which strategies performs bad or too-good-to-be-true and disable them.
-  Strategy(Boost/Handicap)ByPF - EA gives extra boost (or lower it) based on strategy's profit factor (PF).
- (Best/Worse)(Daily/Weekly/Monthly)StrategyMultiplierFactor - EA gives extra boost (or lower it) for strategies based on their PF.
- Con(Wins/Losses)IncreaseFactor - EA gives extra boost based on the consecutive wins or losses. E.g. strategy won 10 trades in a row, make it to calm down by lowering the lot size.
- SmartQueue(Method/Filter) - When MaxOrders (per total, per type or day) is reached and it's not possible to open new trades, new signals are queued for filtering. Once a trading slot is freed (e.g. upon order close or limit increase), EA "decides" which signal from the queue should be processed next based on the method and filter value (e.g. signal from the best performing strategy so far is selected).

So in summary, market conditions can influence performance of strategies, its statistics and EA behaviour in overall.
For example, EA can "learn" that strategy A performing very good and promoting it as best of the week, and increasing its lot size by 50%, and on the other hand, lowering lot size of strategy B by 20% which was performing worse, and so on.

These params were also removed in v2 due to logic refactor (see below). Similar logic can be added soon as part of EA's tasks.

Of course it's not a proper learning, as the trading is still based on the fixed configured user param values. More learning capabilities has been added into EA's VIP versions by utilizing machine learning models.

### v1 to v2 logic refactor

The purpose of the logic refactor from v1 to v2 was to remove all parameters which can affect strategy logic, so each strategy runs in the consistent (aka isolated) environment.

In v1 it was almost impossible to optimize strategy in consistent matter, because there were dozens of EA params affecting trading logic of all strategies at once. So you could optimize strategy, to find out its trading logic breaks after modifying some common params. So basically it was not possible to tell what are the best params for strategy A, because there was too much noise and influence from the other parameters (their trading success was too much depending on these common params, not other way round). And _InitNoOfDaysToWarmUp_, _Boosting_ or _SmartQueue_ params were one of such params affecting the overall results.
Approach for v2 is to make strategies as much isolated as possible (so the optimization per strategy aims towards certain fixed goal and do not break). For that reason, they were moved to separate repositories, so you can run each strategy individually as a tiny EA and find the best setup without external influence.

The mentioned params can be still added in v2, but either per strategy (for the purpose of isolation) or as a task (condition+action). And params for strategy selection (best/worse) can be implemented as meta strategies (a strategy implementing multiple strategies).

**Reply by hrdrq** [reply]
Thank you! The features in v1 should be useful. Are they all removed and v2 is not doing the things?
So the trades should be almost the same whenever the EA starts?
Adding the params into v2's strategies is a plan, or is done already?

**Reply by kenorb** [reply]
> Are they all removed and v2 is not doing the things?

Most of the influential params has been removed in v2 in order to be re-added as per-strategy or tasks.

> So the trades should be almost the same whenever the EA starts?

Yes, it's the same. V2 doesn't have any warm period. Stats and profit factor calculation hasn't been added yet to v2.

> Adding the params into v2's strategies is a plan, or is done already?

Yes, when several pending refactors are done, previous features are going to be re-added one way or another.
